### PR TITLE
Feature: Add double extruder setting

### PR DESCRIPTION
**Pull request by wen-qing-255**

### CODE REVIEW

File: src/app/flux/printing/index.js [hunk 1]
```diff
@@ -313,15 +329,28 @@ export const actions = {
         gcodeLineGroup.position.set(-size.x / 2, -size.y / 2, 0);
     },
 
-    updateDefaultConfigId: (type, defaultId) => (dispatch, getState) => {
+    updateDefaultConfigId: (type, defaultId, direction = 'left') => (dispatch, getState) => {
```
Comment (xiekun1992) [inline src/app/flux/printing/index.js:332]: use constant

File: src/app/flux/printing/index.js [hunk 2]
```diff
@@ -313,15 +329,28 @@ export const actions = {
...
         series = getRealSeries(series);
         let originalConfigId = {};
         if (machineStore.get('defaultConfigId')) {
             originalConfigId = JSON.parse(machineStore.get('defaultConfigId'));
         }
         if (originalConfigId[series]) {
-            originalConfigId[series][type] = defaultId;
+            if (type === PRINTING_MANAGER_TYPE_MATERIAL) {
+                switch (direction) {
+                    case 'left':
```
Comment (xiekun1992) [inline src/app/flux/printing/index.js:342]: use constant

File: src/app/flux/printing/index.js [hunk 3]
```diff
@@ -146,6 +148,7 @@ const INITIAL_STATE = {
     // PrintingManager
     showPrintingManager: false,
     managerDisplayType: PRINTING_MANAGER_TYPE_MATERIAL,
+    materialManagerDirection: 'left',
```
Comment (xiekun1992) [inline src/app/flux/printing/index.js:151]: use constant

**Review by xiekun1992 — CHANGES_REQUESTED** [review]

File: src/app/flux/printing/index.js [hunk 4]
```diff
@@ -1271,6 +1297,25 @@ export const actions = {
...
+        for (const model of modelGroup.selectedModelArray) {
+            const modelItem = lodashFind(models, { modelID: model.modelID });
+            modelItem.extruderConfig = extruderConfig;
+            modelItem.children && modelItem.children.length && modelItem.children.forEach(item => {
+                item.extruderConfig = extruderConfig;
+            });
+        }
+        modelGroup.models = models;
+        dispatch(actions.updateState({
+            modelGroup
```
Comment (jane-rose) [inline src/app/flux/printing/index.js:1312]: update modelGroup is not recommend

File: src/app/flux/project/index.js [hunk 5]
```diff
@@ -117,8 +117,9 @@ export const actions = {
             envObj.coordinateMode = coordinateMode;
             envObj.coordinateSize = coordinateSize;
         } else if (headType === HEAD_PRINTING) {
-            const { defaultMaterialId, defaultQualityId } = editorState;
+            const { defaultMaterialId, defaultMaterialIdRight, defaultQualityId } = editorState;
             envObj.defaultMaterialId = defaultMaterialId;
+            envObj.defaultMaterialIdRight = defaultMaterialIdRight;
```
Comment (jane-rose) [inline src/app/flux/project/index.js:122]: set `defaultMaterialIdRight` a default value

File: src/app/models/ThreeGroup.ts [hunk 6]
```diff
@@ -127,6 +132,20 @@ export default class ThreeGroup extends BaseModel {
...
+            const tempExtruderConfig: ExtruderConfig = Object.assign({}, models[0].extruderConfig as ExtruderConfig);
+            for (const modelItem of models.slice(1)) {
+                const { infill, shell } = (modelItem.extruderConfig as ExtruderConfig);
+                if (infill !== tempExtruderConfig.infill && tempExtruderConfig.infill !== BOTH_EXTRUDER_MAP_NUMBER) {
+                    tempExtruderConfig.infill = BOTH_EXTRUDER_MAP_NUMBER;
+                }
+                if (shell !== tempExtruderConfig.shell && tempExtruderConfig.shell !== BOTH_EXTRUDER_MAP_NUMBER) {
+                    tempExtruderConfig.shell = BOTH_EXTRUDER_MAP_NUMBER;
+                }
+                if (tempExtruderConfig.infill === BOTH_EXTRUDER_MAP_NUMBER && tempExtruderConfig.shell === BOTH_EXTRUDER_MAP_NUMBER) {
```
Comment (jane-rose) [inline src/app/models/ThreeGroup.ts:144]: What does this mean? Add code comments?

**Review by jane-rose — APPROVED** [review]

**Review by xiekun1992 — APPROVED** [review]

**Pull request merged**